File: packages/dart/test/interpolation_test.dart
import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

void main() {
  group('detectInterpolationDelimiters', () {
    test('reads each pair it knows', () {
      expect(
        detectInterpolationDelimiters('Hello {name}'),
        const Chki18nDelimiters(prefix: '{', suffix: '}'),
      );
      expect(
        detectInterpolationDelimiters('Hello [[name]]'),
        const Chki18nDelimiters(prefix: '[[', suffix: ']]'),
      );
      expect(
        detectInterpolationDelimiters('Hello ((name))'),
        const Chki18nDelimiters(prefix: '((', suffix: '))'),
      );
      expect(
        detectInterpolationDelimiters('Hello <name>'),
        const Chki18nDelimiters(prefix: '<', suffix: '>'),
      );
    });

    test('reads a doubled pair as itself, not as its single form', () {
      expect(
        detectInterpolationDelimiters('Hello {{name}}'),
        const Chki18nDelimiters(prefix: '{{', suffix: '}}'),
      );
    });

    test('allows the spacing a style may put inside the delimiters', () {
      expect(
        detectInterpolationDelimiters('Hello {{ name }}'),
        const Chki18nDelimiters(prefix: '{{', suffix: '}}'),
      );
    });

    test('is not fooled by the punctuation of the JSON holding the text', () {
      expect(detectInterpolationDelimiters('{"desc":{"hello":"Hello"}}'), isNull);
      expect(detectInterpolationDelimiters('{\n\t"list": ["a", "b"]\n}'), isNull);
    });

    test('answers with the first pair it believes when a text mixes two', () {
      expect(
        detectInterpolationDelimiters('{{a}} and [[b]]'),
        const Chki18nDelimiters(prefix: '{{', suffix: '}}'),
      );
    });

    test('has nothing to say about a text with no placeholder', () {
      expect(detectInterpolationDelimiters('Hello there'), isNull);
      expect(detectInterpolationDelimiters(''), isNull);
    });

    test('answers with one of the pairs it publishes', () {
      expect(interpolationDelimiters, contains(detectInterpolationDelimiters('Hello {name}')));
    });
  });

  group('the delimiters a scan saw', () {
    test('reports what the files it read are written with', () async {
      final session = await loadTranslations(path: 'test/samples/locales-no-issue');

      expect(session.detectedInterpolation, const Chki18nDelimiters(prefix: '{', suffix: '}'));
      // A guess about the files, not the setting the checks ran with.
      expect(session.options.interpolationPrefix, '{');
    });

    test('says nothing about files that hold no placeholder', () async {
      final session = await loadTranslations(path: 'test/samples/excluded-files');

      expect(session.detectedInterpolation, isNull);
    });

    test('says nothing when there was no directory to read', () async {
      final session = await loadTranslations();

      expect(session.detectedInterpolation, isNull);
    });
  });
}

File: packages/javascript/test/interpolation.test.ts
import assert from 'node:assert';
import { join } from 'node:path';
import { describe, it } from 'node:test';
import {
	detectInterpolationDelimiters,
	INTERPOLATION_DELIMITERS,
	loadTranslations
} from '../dist/index.js';

describe('detectInterpolationDelimiters', () => {
	it('reads each pair it knows', () => {
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello {name}'), {
			prefix: '{',
			suffix: '}'
		});
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello [[name]]'), {
			prefix: '[[',
			suffix: ']]'
		});
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello ((name))'), {
			prefix: '((',
			suffix: '))'
		});
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello <name>'), {
			prefix: '<',
			suffix: '>'
		});
	});

	it('reads a doubled pair as itself, not as its single form', () => {
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello {{name}}'), {
			prefix: '{{',
			suffix: '}}'
		});
	});

	it('allows the spacing a style may put inside the delimiters', () => {
		assert.deepStrictEqual(detectInterpolationDelimiters('Hello {{ name }}'), {
			prefix: '{{',
			suffix: '}}'
		});
	});

	it('is not fooled by the punctuation of the JSON holding the text', () => {
		assert.strictEqual(detectInterpolationDelimiters('{"desc":{"hello":"Hello"}}'), null);
		assert.strictEqual(detectInterpolationDelimiters('{\n\t"list": ["a", "b"]\n}'), null);
	});

	it('answers with the first pair it believes when a text mixes two', () => {
		assert.deepStrictEqual(detectInterpolationDelimiters('{{a}} and [[b]]'), {
			prefix: '{{',
			suffix: '}}'
		});
	});

	it('has nothing to say about a text with no placeholder', () => {
		assert.strictEqual(detectInterpolationDelimiters('Hello there'), null);
		assert.strictEqual(detectInterpolationDelimiters(''), null);
	});

	it('answers with one of the pairs it publishes', () => {
		const found = detectInterpolationDelimiters('Hello {name}');

		assert.ok(
			INTERPOLATION_DELIMITERS.some(
				(pair) => pair.prefix === found?.prefix && pair.suffix === found?.suffix
			)
		);
	});
});

describe('the delimiters a scan saw', () => {
	it('reports what the files it read are written with', async () => {
		const session = await loadTranslations(join('test', 'samples', 'locales-no-issue'));

		assert.deepStrictEqual(session.detectedInterpolation, { prefix: '{', suffix: '}' });
		// A guess about the files, not the setting the checks ran with.
		assert.strictEqual(session.options.interpolationPrefix, '{');
	});

	it('says nothing about files that hold no placeholder', async () => {
		const session = await loadTranslations(join('test', 'samples', 'excluded-files'));

		assert.strictEqual(session.detectedInterpolation, null);
	});

	it('says nothing when there was no directory to read', async () => {
		const session = await loadTranslations();

		assert.strictEqual(session.detectedInterpolation, null);
	});
});
